keyboards/hub16/keymaps/default/keymap.c

```diff
@@ -0,0 +1,143 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include QMK_KEYBOARD_H
+
+// Allow reading of encoder switch
+#define readInput(port, pin) (((*(&port - 2)) & (1 << pin)) ? 1 : 0)
+#define SWITCH_1 PF7
+#define SWITCH_1_PORT PORTF
+#define SWITCH_2 PD7
+#define SWITCH_2_PORT PORTD
+
+// Function key we are 'wrapping' usual key presses in
+#define KC_WRAP KC_F24
+
+// Tap Dance Declarations
+enum { TD_TO_LED = 0, TD_TO_DEFAULT = 1 };
+
+qk_tap_dance_action_t tap_dance_actions[] = {
+    // Tap once for HID, twice to toggle layer 1
+    [TD_TO_LED]     = ACTION_TAP_DANCE_DUAL_ROLE(_______, 1),
+    [TD_TO_DEFAULT] = ACTION_TAP_DANCE_DUAL_ROLE(_______, 0)};
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+  [0] = LAYOUT( /* Base */
+    KC_A, KC_B, KC_C, KC_D, \
+    KC_E, KC_F, KC_G, KC_H, \
+    KC_I, KC_J, KC_K, KC_L, \
+    KC_M, KC_N, KC_O, TD(TD_TO_LED) \
+  ),
+
+  [1] = LAYOUT( /* LED Control */
+    _______, RGB_MOD, RGB_RMOD, RGB_TOG, \
+    RGB_VAD, RGB_VAI, RGB_HUD, RGB_HUI, \
+    RGB_SAD, RGB_SAI, _______, _______, \
+    _______, _______, RESET, TD(TD_TO_DEFAULT) \
+  ),
+};
+
+// Keyboard is setup to 'warp' the pressed key with F24,
+// allowing for easy differentiation from a real keyboard.
+void encoder_update_user(uint8_t index, bool clockwise) {
+    if (index == 0) { /* Left Encoder */
+        if (clockwise) {
+            register_code(KC_WRAP);
+            tap_code(KC_R);
+            unregister_code(KC_WRAP);
+        } else {
+            register_code(KC_WRAP);
+            tap_code(KC_Q);
+            unregister_code(KC_WRAP);
+        }
+    } else if (index == 1) { /* Right Encoder */
+        if (clockwise) {
+            register_code(KC_WRAP);
+            tap_code(KC_U);
+            unregister_code(KC_WRAP);
+        } else {
+            register_code(KC_WRAP);
+            tap_code(KC_T);
+            unregister_code(KC_WRAP);
+        }
+    }
+}
+
+// Below stolen from TaranVH (https://github.com/TaranVH/2nd-keyboard/blob/master/HASU_USB/F24/keymap.c)
+// Shoutout to drashna on the QMK discord for basically writing this for me.... :P
+
+bool process_record_user(uint16_t keycode, keyrecord_t *record) {
+    static uint8_t f24_tracker;
+    switch (keycode) {
+        case KC_A ... KC_F23:
+        case KC_EXECUTE ... KC_EXSEL:
+            if (record->event.pressed) {
+                register_code(KC_WRAP);
+                f24_tracker++;
+                register_code(keycode);
+            } else {
+                unregister_code(keycode);
+                f24_tracker--;
+                if (!f24_tracker) {
+                    unregister_code(KC_WRAP);
+                }
+            }
+            return false;
+            break;
+    }
+    return true;
+}
+
+void matrix_init_user(void) {
+    // Set encoder switches as inputs
+    setPinInput(SWITCH_1);
+    setPinInput(SWITCH_2);
+}
+
+void matrix_scan_user(void) {
```

Comment:
It ma ybe better/simpler to make a custom matrix for this board, then. 

That way, you don't need to have this code here, and it can be in the normal LAYOUT macro.

Reply:
Thanks, I hadn't thought of a custom matrix, not sure how.. 
Got it recognising the key presses, however they would appear to be stuck down indefinitely afterwards. 
If possible I'd like to merge it as is, however understand if it needs to be resolved prior. If so do you know of any examples which I may be able to reference? Thanks!

Reply:
Basically, look for any keyboard using `CUSTOM_MATRIX = yes`.  You need to include your own matrix, but you can copy the default one in `/quantum/` and modify it from there.

Reply:
Thanks for the pointer, got it working! Not elegant but better than before.